blockchain/standalone/pow.go

```diff
@@ -182,22 +182,217 @@ func CheckProofOfWorkRange(difficultyBits uint32, powLimit *big.Int) error {
+	//
+	// In particular, the goal cubic polynomial approximation over the interval
+	// 0 <= x < 1 is:
+	//
+	//   2^x ~= 1 + 0.695502049712533x + 0.2262697964x^2 + 0.0782318x^3
+	//
+	// This approximation provides an absolute error margin < 0.013% over the
+	// aforementioned interval of [0,1) which is well under the 0.1% error
+	// margin needed for good results.  Note that since the input domain is not
+	// constrained to that interval, the exponent is decomposed into an integer
+	// part, n, and a fractional part, f, such that f is in the desired range of
+	// [0,1).  By exponent rules 2^(n + f) = 2^n * 2^f, so the strategy is to
+	// calculate the result by applying the cubic polynomial approximation to
+	// the fractional part and using the fact that multiplying by 2^n is
+	// equivalent to an arithmetic left or right shift depending on the sign.
+	//
+	// In other words, start by calculating the exponent (x) using 64.16 fixed
+	// point and decompose it into integer (n) and fractional (f) parts as
+	// follows:
+	//
+	//       2^16 * (Δt - Δh*Ib)   (Δt - Δh*Ib) << 16
+	//   x = ------------------- = ------------------
+	//            halfLife              halfLife
+	//
+	//        x
+	//   n = ---- = x >> 16
+	//       2^16
+	//
+	//   f = x (mod 2^16) = x & 0xffff
+	//
+	// The use of 64.16 fixed point for the exponent means both the integer (n)
+	// and fractional (f) parts have an additional factor of 2^16.  Since the
+	// fractional part of the exponent is cubed in the polynomial approximation
+	// and (2^16)^3 = 2^48, the addition step in the approximation is internally
+	// performed used 16.48 fixed point to compensate.
+	//
+	// In other words, the fixed point formulation of the goal cubic polynomial
+	// approximation for the fractional part is:
+	//
+	//                 195766423245049*f + 971821376*f^2 + 5127*f^3 + 2^47
+	//   2^f ~= 2^16 + ---------------------------------------------------
+	//                                          2^48
+	//
+	// Finally, the final target difficulty is calculated using x.16 fixed point
+	// and then clamped to the valid range as follows:
+	//
+	//              startDiff * 2^f * 2^n
+	//   nextDiff = ---------------------
+	//                       2^16
+	//
+	//   nextDiff = min(max(nextDiff, 1), powLimit)
+	//
+	// NOTE: The division by the half life uses Quo instead of Div because it
+	// must be truncated division (which is truncated towards zero as Quo
+	// implements) as opposed to the Euclidean division that Div implements.
+	idealTimeDelta := heightDelta * targetSecsPerBlock
+	exponentBig := big.NewInt(timeDelta - idealTimeDelta)
+	exponentBig.Lsh(exponentBig, 16)
+	exponentBig.Quo(exponentBig, big.NewInt(halfLife))
+
+	// Decompose the exponent into integer and fractional parts.  Since the
+	// exponent is using 64.16 fixed point, the bottom 16 bits are the
+	// fractional part and the integer part is the exponent right shifted by 16.
```

Comment:
Since the exponent is signed, this should explicitly say it is **arithmetic** right shifted because a logical right shift will not produce the correct result for negative values.

Reply:
> commit "chaingen: Add ASERT difficulty algorithm support. " has
> 
> > This adds support for generating blocks **the** require the new difficulty

Fixed.

Reply:
Added.